### ISSUE TITLE
Any  interest in adding ML Perf  Llama 3 8B to TorchTitan models ?

### ISSUE THREAD

**Issue by githubsgi** [pytorch/torchtitan#2046]
It will be great to have MLPerf LLama 3 pre-training working OOB with TorchTitan,  Here are some references on that .

[MLPerf Training Adds Llama 3.1 8B Benchmark](https://mlcommons.org/2025/10/training-llama-3-1-8b/)

[small_llm_pretraining/nemo](https://github.com/mlcommons/training/tree/master/small_llm_pretraining/nemo)

**Comment by githubsgi** [comment]
@tianyu-l , @wconstab , any thoughts on this ?  Having such a model OOB  would greatly  facilitates quick validation and comparison or full HW and SW stack.

**Comment by tianyu-l** [comment]
> Having such a model OOB would greatly facilitates quick validation and comparison or full HW and SW stack.

sorry, could you tell a more complete story? Not clear to me what's going on.

**Comment by githubsgi** [comment]
@tianyu-l , a quick summary of MLPerf follows. 

MLCommons is the world’s leader in AI benchmarking. An open engineering consortium supported by over 125 members and affiliates, MLCommons has a proven record of bringing together academia, industry, and civil society to measure and improve AI. MLPerf is an industry-standard benchmark suite from MLCommons for measuring the performance of machine learning (ML) hardware, software, and cloud platforms. It provides standardized, apples-to-apples comparisons across different ML tasks, covering both the training of AI models and the inference (or running) of those models on a wide range of systems, from data centers to mobile devices. The benchmarks are a result of a consortium of AI industry leaders, researchers, and academics.

MLPerf Training v5.1 introduces a benchmark based on Meta’s Llama 3.1 8B, replacing BERT with a modern model that can still run on single-node systems. This new pretraining benchmark based on Meta’s Llama 3.1 8B model brings modern LLM pretraining evaluation within reach of more organizations. It was intentionally designed to be easy to set up and run on small to moderately sized computational resources. By requiring only a single node, using a subset of the C4 dataset, and starting from random weights, it lowers the barrier to entry while maintaining relevance to current AI development practices.

**Comment by githubsgi** [comment]
@tianyu-l ,  does that give you enough context about MLPerf ?   I am only proposing support for Llama3 8B MLPerf pre-training case - not any of the other MLPerf case at this point.

**Comment by tianyu-l** [comment]
I still don't know what's the point of supporting MLPerf version of Llama3 8B, how's it different from current Llama 3 8B. Did you mention it anywhere?

**Comment by githubsgi** [comment]
@tianyu-l ,  I did try it out. Works with some changes. 

1.  The c4 en dataset , training and validation,  is prepared  slightly differently.  
2.  The dataset has to be added to the file "text_datasets.py" as described in   [datasets.md](https://github.com/pytorch/torchtitan/blob/main/docs/datasets.md)
3. [Log perplexity](https://github.com/NVIDIA-NeMo/NeMo/blob/fb633b409d22179caaea35cc2d7a488c15870fe0/nemo/collections/common/metrics/perplexity.py#L22) ( target = 3.3) needs to be added to track  accuracy.

**Comment by githubsgi** [comment]
@tianyu-l ,  please let me know your thoughts. I am  planning to do a PR.

**Comment by wconstab** [comment]
IIUC what you're proposing is to change some things in Torchtitan so that you can directly run the MLPerf llama experiment using unchanged torchtitan.  Is that right?  Well, it depends what the changes are!

Can you give more information about the proposed changes before you submit a PR?  (just to save yourself time in case we don't like the changes).  

For example
> The c4 en dataset , training and validation, is prepared slightly differently.

What is the difference?  What is the motivation?

...

**Comment by tianyu-l** [comment]
@githubsgi I tried asking what's the motivation and benefit but I have NOT got an answer from you.

**Comment by githubsgi** [comment]
@tianyu-l ,  please see if this makes sense to you.  As you pointed out, integrating the MLPerf Llama3 8B model into TorchTitan  would allow  running a  widely used performance benchmark out of box (OOB). TorchTitan's hardware agnostic design enables new accelerators to be onboarded with minimal friction, making it an ideal testbed for emerging hardware. By incorporating the MLPerf Llama3 8B benchmark, we create a  OOB reference point that allows new accelerators to be rapidly validated  against a well-established industry benchmark quickly.

The benefits extend beyond just hardware validation. MLPerf Llama3 8B submissions provide a comprehensive ecosystem of reference results, detailed logs, convergence metrics, and performance baselines across multiple accelerators from multiple vendors. This wealth of data serves as a crucial calibration tool for new accelerators, enabling engineers to quickly identify performance gaps, correctness issues, optimization opportunities, etc.

**Comment by wconstab** [comment]
I think what would help me is to understand the finer grain proposal and its tradeoffs.  For example, you mentioned you want to change something about the dataset.  Why, and with what implications?   That level of detail for each change would be easier to discuss.

**Comment by githubsgi** [comment]
@wconstab ,   there  are essentially 2 minor differences that I can see  as far as data prep goes .  1.  A download script is provided by MLCommons for downloading the C4 dataset  from their  site. 2.   Another script is provided to merge the  training set into 8 files  (e.g. c4-train.en_0.json.gz )  and validation set into 1 file  ( e.g. c4-validation.en.json.gz ) .    There is a third step for  converting to Megatron formats  - I do not think these formats are supported  and relevant (?).  These 3 steps are described [here](https://github.com/mlcommons/training/tree/master/small_llm_pretraining/nemo#raw-data-downloading-optional).

**Comment by githubsgi** [comment]
@wconstab , any comment ? 

Also  looked into the loss function for Llama3 8B. It is Cross Entropy,  hence perplexity = 2**loss. So, a perplexity of 3.3  is 1.72246602447  in loss.  

Hence , if the above is correct,  then the only thing that is needed  for an MLPerf compatible run is  the data prep !

**Comment by githubsgi** [comment]
@wconstab and @tianyu-l , any comments on the above ?  Do you see benefits in adding a condensed version of this issue in the README page ?

**Comment by githubsgi** [comment]
Looks like I missed  the dataset add to torchtitan/torchtitan/hf_datasets/text_datasets.py

```
def _load_c4_mlperf_dataset(dataset_path: str, split: str):
    """Load C4 dataset with default configuration."""
    return load_dataset(dataset_path, name="", split=split, streaming=True)
.
.
.
DATASETS = {
.
.
.
    "c4_mlperf": DatasetConfig(
        path="mlperf/c4",
        loader=partial(_load_c4_mlperf_dataset, split="train"),
        sample_processor=_process_c4_text,
    ),
    "c4_mlperf_validation": DatasetConfig(
        path="mlperf/c4",
        loader=partial(_load_c4_mlperf_dataset, split="validation"),
        sample_processor=_process_c4_text,
    ),
}

```

**Comment by wconstab** [comment]
I still don't understand what's being proposed for the changes to c4 dataset in torchtitan. You mentioned providing a download script, generating a specific number of files, and megatron compatibility.

I'm not sure why I care how the files are chunked or whether it's compatible with megatron- you could presumably follow megatron's requirements when running megatron right? Is the request that once you follow torchtitan setup procedure then megatron also runs out of the box?

**Comment by githubsgi** [comment]
@wconstab, you are probably aware of what  [MLPerf ](https://mlcommons.org/2025/10/training-llama-3-1-8b/) is and its importance to hardware/silicon vendors. The goal is to use TorchTitan, not Megatron, to run the MLPerf SLM test.  No asking MLPerf  to use TorchTitan in lieu of or in addition to Megatron.  This is for informal engineering comparison of  new silicon/systems as MLPerf  has tons of performance data.

**Comment by wconstab** [comment]
hm, so we already have support for c4 dataset, and you can use it to run torchtitan. Can you clarify what is not good enough about the existing c4 support? or is it actually fine and you just didn't understand what torchtitan offered at first?

**Comment by githubsgi** [comment]
@wconstab ,  they provide their own C4 dataset and does some post-processing ( through scripts they provide). The post-processed data looks better suited  from storage I/O  related bottlenecks perspective, compared to the plain vanilla  C4 dataset. Due to small number of large files vs large number of small files.

**Comment by wconstab** [comment]
is the dataset the same in terms of which examples it provides for training? and it just offers a throughput optimization for training based on reorganizing the files on disk?  I think we could consider landing that in torchtitan IF its not affecting training numerics and strictly optimizing the system behavior for dataloading.  However if it affects the numerics of training, landing it is more tricky since it could invalidate a bunch of pre-existing experiments, and i'm not sure if that'd be worth it.

**Comment by githubsgi** [comment]
@wconstab , here are  some snippets from [MLPerf github ](https://github.com/mlcommons/training/tree/master/small_llm_pretraining/nemo#raw-data-downloading-optional) 


> We use [AllenAI C4](https://huggingface.co/datasets/allenai/c4) dataset for this benchmark. The original zipped json.gz files can be downloaded by following AllenAI C4's instruction, and you can download our zipped customized validation dataset from the MLCommons S3 ...
> 
> After downloading, run the following command to process them to zip them into .gz format before running the data preprocessing.
> 
> Run the following commands to merge all 1024 training files into 8 json.gz files, all 8 validation files into a single json.gz file, as well as generate our customized validation dataset.
> 
> Training and test data separation
> We use the default split from the C4 dataset. This means that we use c4-train.<x>-of-01024.json.gz files (where 768 <= x <= 1023) for training, and we use our customized c4-validation-91205-samples.en.json.gz, which contains the first 91205 samples from the unshuffled C4 validation dataset, for evaluation.
> 
> Notice here that we are using the first 1024 sequences (8,388,608 tokens) from the validation dataset to perform the validation. According to our experiments, the first 91205 samples from the unshuffled C4 dataset yields 47,186,855 tokens, which is the smallest amount of samples needed to yield 47,185,920 tokens. Thus, we have chosen the first 91205 samples as our validation dataset.
> 
> Training data order
> We randomly shuffle the last 256 of 1024 shards for the benchmarking area.
> 
> Test data order
> We use the first 1024 sequences in the validation dataset for validation. We do not shuffle the validation dataset.
